### PR TITLE
kubeadm: Move .NodeName and .CRISocket to a common sub-struct

### LINKED ISSUE #847: Make it possible to configure the kubelet via the v1alpha2 Config file
When kubernetes/kubernetes#63887 merges, we are able to set various flags for the kubelet's CLI. These flags should really be kept to a minimum, but for some things like hostname-override, taints, labels, node-ip etc. this is the only option. The same for very host/OS-specific configuration like [`--cgroup-driver`](https://github.com/kubernetes/kubeadm/issues/844) and [`--resolv-conf`](https://github.com/kubernetes/kubeadm/issues/845).

The proposal here is to allow the user to specify these extraargs via the Config file:

```go
Labels []string `json:"labels,omitempty"`

Taints []v1.Taint `json:"taints,omitempty"`

KubeletExtraArgs map[string]string `json:"kubeletExtraArgs,omitempty"`
```

(these would not be top-level, but nested inside some struct under the `MasterConfiguration` initially.)

The `KubeletExtraArgs` string-string map would work as any other control plane extraargs map we have.
`if Taints == nil` at default time, the default master taint is gonna be applied. If `taints: {}` in the YAML file, no taints are gonna be added, which equals the behavior of `NoTaintMaster` which we can then remove.
`if Labels == nil` at default time, the default master label is gonna be applied. If `labels: {}` in the YAML file, no labels are gonna be added.
This also works for `kubeadm join`, apart from the master taint/label logic.
This will fix numerous issues like #209 #202 #203 + more.

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Regroups some common fields for `kubeadm init` and `kubeadm join` only used for the initial node registration.
Lets the user specify ExtraArgs to the kubelet.
Now also runs the dynamic env file creation for `kubeadm join`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/847
Follows-up https://github.com/kubernetes/kubernetes/pull/63887
Related to kubernetes/kubeadm#822

**Special notes for your reviewer**: WIP, but please review so we can finalize the direction of the PR

**Release note**:

```release-note
[action required] `.NodeName` and `.CRISocket` in the `MasterConfiguration` and `NodeConfiguration` v1alpha1 API objects are now `.NodeRegistration.Name` and `.NodeRegistration.CRISocket` respectively in the v1alpha2 API. The `.NoTaintMaster` field has been removed in the v1alpha2 API.
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @liztio

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 1]
```diff
@@ -136,6 +129,27 @@ type API struct {
...
+	CRISocket string
+
+	// Taints specifies the taints the Node API object should be registered with. If this field is unset, i.e. nil, in the `kubeadm init` process
+	// it will be defaulted to []v1.Taint{'node-role.kubernetes.io/master=""'}. If you don't want to taint your master node, set this field to an
+	// empty slice, i.e. `taints: {}` in the YAML file. This field is solely used for Node registration.
+	Taints []v1.Taint
+
+	// ExtraArgs passes through extra arguments to the kubelet. The arguments here are passed to the kubelet command line via the environment file
+	// kubeadm writes at runtime for the kubelet to source.
+	ExtraArgs map[string]string
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: Given that this is the Kubelet ExtraArgs it's not entirely clear from a developer UX.  

For example: 
NodeRegistrationOptions.ExtraArgs

vs. 
KubeletRegistrationOptions.ExtraArgs or 
NodeRegistrationOptions.KubeletExtraArgs.

File: cmd/kubeadm/app/util/apiclient/idempotency.go [hunk 2]
```diff
@@ -186,3 +192,49 @@ func CreateOrUpdateClusterRoleBinding(client clientset.Interface, clusterRoleBin
 	}
 	return nil
 }
+
+// PatchNode tries to patch a node using the following client, executing patchFn for the actual mutating logic
+func PatchNode(client clientset.Interface, nodeName string, patchFn func(*v1.Node)) error {
+	// Loop on every falsy return. Return with an error if raised. Exit successfully if true is returned.
```
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: false return.

File: cmd/kubeadm/app/util/apiclient/idempotency.go [hunk 3]
```diff
@@ -186,3 +192,49 @@ func CreateOrUpdateClusterRoleBinding(client clientset.Interface, clusterRoleBin
 	}
 	return nil
 }
+
+// PatchNode tries to patch a node using the following client, executing patchFn for the actual mutating logic
+func PatchNode(client clientset.Interface, nodeName string, patchFn func(*v1.Node)) error {
```
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: Do you patch b/c dynamic kubelet config is not default?

**Review by timothysc — COMMENTED** [review]
Minor nits then lgtm

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 4]
```diff
@@ -136,6 +129,27 @@ type API struct {
 	BindPort int32
 }
 
+// NodeRegistrationOptions holds fields that relate to registering a new master or node to the cluster, either via "kubeadm init" or "kubeadm join"
+type NodeRegistrationOptions struct {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: If we are considering convergence towards cluster API, the [corresponding object](https://github.com/kubernetes/kube-deploy/blob/5d7e0ecaec541c561f6e18cb4500c11554cfa406/cluster-api/pkg/apis/cluster/v1alpha1/machine_types.go#L51) is named `machineSpec`. 
WDYT? it is too early for starting to unify naming?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 5]
```diff
@@ -56,7 +56,8 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
 	allErrs = append(allErrs, ValidateCertSANs(c.Etcd.ServerCertSANs, field.NewPath("etcd").Child("serverCertSANs"))...)
 	allErrs = append(allErrs, ValidateCertSANs(c.Etcd.PeerCertSANs, field.NewPath("etcd").Child("peerCertSANs"))...)
 	allErrs = append(allErrs, ValidateAbsolutePath(c.CertificatesDir, field.NewPath("certificatesDir"))...)
-	allErrs = append(allErrs, ValidateNodeName(c.NodeName, field.NewPath("nodeName"))...)
+	allErrs = append(allErrs, ValidateNodeName(c.NodeRegistration.Name, field.NewPath("nodeRegistration").Child("name"))...)
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: What about creating `validateNodeRegistrationOptions` ?

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 6]
```diff
@@ -136,6 +129,27 @@ type API struct {
...
+
+	// CRISocket is used to retrieve container runtime info. This information will be annotated to the Node API object, for later re-use
+	CRISocket string
+
+	// Taints specifies the taints the Node API object should be registered with. If this field is unset, i.e. nil, in the `kubeadm init` process
+	// it will be defaulted to []v1.Taint{'node-role.kubernetes.io/master=""'}. If you don't want to taint your master node, set this field to an
+	// empty slice, i.e. `taints: {}` in the YAML file. This field is solely used for Node registration.
+	Taints []v1.Taint
+
+	// ExtraArgs passes through extra arguments to the kubelet. The arguments here are passed to the kubelet command line via the environment file
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: What about adding a note that explain difference between  extra args and KubeletConfiguration (machine specific vs cluster wide)

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 5 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: This is WIP, thanks for keeping me honest fixing this as well

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 6 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Very good point :+1:

File: cmd/kubeadm/app/util/apiclient/idempotency.go (hunk 2 above)
Comment (luxas) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: done

File: cmd/kubeadm/app/util/apiclient/idempotency.go (hunk 3 above)
Comment (luxas) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: No this is used when e.g. setting the master label/taint, and generically modifying the Node object.

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: I'll change this after we've discussed the naming with @fabriziopandini in the SIG meeting

**Review by timothysc — APPROVED** [review]
/lgtm 

Please open an issue to determine long term plan for translating <> machinespec.

**Pull request merged**